src/User/User.php

```diff
@@ -111,6 +113,49 @@ public function setIdentity(IdentityInterface $identity): void
         $this->identity = $identity;
     }
 
+    /**
+     * Return the auth key value
+     *
+     * @return string Auth key value
+     */
+    public function getAuthKey(): string
+    {
+        return 'ABCD1234';
+    }
+
+    /**
+     * Validate auth key
+     *
+     * @param String $authKey Auth key to validate
+     * @return bool True if is valid
+     */
+    public function validateAuthKey($authKey): bool
+    {
+        return $authKey === 'ABCD1234';
+    }
+
+    /**
+     * Sends an identity cookie.
+     *
+     * @param IdentityInterface $identity
+     * @param int $duration number of seconds that the user can remain in logged-in status.
+     */
+    protected function sendIdentityCookie(IdentityInterface $identity, int $duration): void
+    {
+        $data = json_encode(
+            [
+                $identity->getId(),
+                $this->getAuthKey(),
+                $duration,
+            ],
+            JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE
+        );
+
+        $cookieIdentity = (new Cookie('remember', $data))->expireAt(time() + $duration);
+        $response = new Response();
```

Comment:
If you need to create new Response instance then you should use ResponseFactory from PSR

Reply:
Can you shared me an example please?. Because I can't find the namespace. Thanks!

Reply:
for example you can explore [#236](https://github.com/yiisoft/yii-web/pull/236/files)

Reply:
@roxblnfk thanks. I have a question. In the class user I have the sendIdentityCookie method, this is called via method login. In the same class. I have the responseFactory in AutoLoginMiddleware. Is it a good idea to pass the response in the login method as a parameter?

Reply:
Wow. I think the User class should not handle either the Response instance or the Request instance.

Reply:
In addition, if you do not return a Response object, then after executing the `sendIdentityCookie ()` method, the object will be destroyed

Reply:
I think the same, that's why I asked the question. I understand if I do not return a Response the cookie won't the save. But I saw yii-2 and I saw this method Yii::$app->getResponse()->getCookies()->add($cookie);  Is there an equivalent?

Reply:
You can subscribe to `AfterLogin` and `AfterLogout` events.

Reply:
@roxblnfk ok, I'm getting closer, but it's still not clear to me. How can I subscribe?. I also don't understand the logic of these events, classes are just getters. Thank you very much.

Reply:
Sorry for my absence. As it turned out, now in yii3 you can’t subscribe to events and unsubscribe from it in runtime.
This imposes a limitation on the use of the event system in middleware.
You can try another option: check the status of the user before and after receiving a Response from the next middleware. Based on the status, set cookies in the Response object

Reply:
@samdark ahh ok I understand that in yii3 not is possible subscribe. But this "check the status of the user before and after receiving a Response from the next middleware. Based on the status, set cookies in the Response object" Not is clear for me. Is there a little example for me to understand?. Thanks!

Reply:
Simple example:
```php
/** @var \Yiisoft\Yii\Web\User\User $user */
if ($user->isGuest())
{
    // read cookies
}
// continue pipeline
$response = $handler->handle($request);
if (!$user->isGuest())
{
    // write cookies to $response
}
return $response;
```